### PR TITLE
storage: use only PF4 components and stop including PF3 stylesheets

### PR DESCRIPTION
**Pull request by KKoukiou**
This commit:
* replaces icons with components from react-icons, also makes sure these
  are accessible
* replaces custom lists with DataLists component
* replaces all dialog fields like TextInputs, Checkboxes, Radio buttons with PF4
  components
* replaces custom dropdowns with Dropdown component

Todo:

- [x] The journal has lost the markup for repated entries, https://github.com/cockpit-project/cockpit/pull/15261#issuecomment-783273679
- [x] Spinner and progress text are too close together in dialogs, https://github.com/cockpit-project/cockpit/pull/15261#issuecomment-783277169
- [x] "Encryption options" dialog repeats itself, https://github.com/cockpit-project/cockpit/pull/15261#issuecomment-783286187

### CODE REVIEW

File: pkg/storaged/things-panel.jsx
Comment (mvollmer) [inline pkg/storaged/things-panel.jsx:69]: What does the `filter` do here?

Comment (KKoukiou) [reply pkg/storaged/things-panel.jsx:69]: Some items here are sometimes null, this PF4 component get's really upset about null children.

File: pkg/storaged/dialog.jsx
Comment (mvollmer) [inline pkg/storaged/dialog.jsx:466]: I don't understand the addition of `validated` here.  I don't think any caller actually passes in a value for it, and I dount that the PF4 TextInput has a `validated` property.

Comment (KKoukiou) [reply pkg/storaged/dialog.jsx:466]: Oh it does https://www.patternfly.org/v4/components/text-input#props

And I pass it to the children here https://github.com/cockpit-project/cockpit/pull/15261/files/7830bd747feb82212dad66d45976d56407b5c2d0#diff-5b932241940b997840c4960d1bda3e884d3ba162f9aa007b761920aa61fdd245R257

File: pkg/storaged/things-panel.jsx
Comment (mvollmer) [reply pkg/storaged/things-panel.jsx:69]: Ah, makes sense. I had confused `filter` with `map`... d'oh.  What about writing `filter(item => item !== null)`?

Comment (KKoukiou) [reply pkg/storaged/things-panel.jsx:69]: Sure I will replace on next repush.

File: pkg/storaged/dialog.jsx
Comment (mvollmer) [reply pkg/storaged/dialog.jsx:466]: > Oh it does https://www.patternfly.org/v4/components/text-input#props

Ohh!  Once again I wasn't able to find any proper documentation. Google send me here: https://www.patternfly.org/2020.04/documentation/react/components/textinput which has no "props" section.

Comment (mvollmer) [reply pkg/storaged/dialog.jsx:466]: > And I pass it to the children here

I think that doesn't actually go through the `TextInput` function, and will directly work on the result of the `render` function returned by `TextInput`.  I'll check this out more carefully.

(The things like `TextInput` are not React components, but I realize they look awefuly close.)

Comment (mvollmer) [reply pkg/storaged/dialog.jsx]: I have pushed a FIXUP with how I think it needs to be done.  In the case of `SelectOne`, the React.clone approach would stick the "validated" prop on the `div` instead of on the `FormSelect` where it needs to be.

**Comment by KKoukiou** [comment]
@mvollmer, can I get another review on this one? There is also one test failing, TestStorageUnused.testUnused, which I can't figure out what wrong.

**Comment by mvollmer** [comment]
There is now a gap between the dropdown button and the dropdown menu:

image

Previously:

image

**Comment by KKoukiou** [comment]
> There is now a gap between the dropdown button and the dropdown menu:
> 
> image
> 
> Previously:
> 
> image

The issue here is that PF4 does not genuinely support buttons that expand dropdowns. I abused the Dropdown component to make this happen, so I don't expect that's ideal.

**Comment by mvollmer** [comment]
RAID dialog:

image

Previously:

image

There is now a blue border all around a focused text input field. Is that intentional?

The disk selection should have the block device names right aligned.  The borders have changed.

**Comment by mvollmer** [comment]
VDO dialog:

image

Previously:

image

The (i) bubbles are too far away from the text that they belong to.

**Comment by mvollmer** [comment]
Text field of size sliders look differently now when focused:

image

Previously:

image

**Comment by mvollmer** [comment]
Groups of checkboxes in dialogs look cramped and are not aligned with their label:

image

Previously:

image

**Comment by mvollmer** [comment]
Dropdowns have a blue focus rings as well now:

image

but sometimes a blue focus line:

image

**Comment by mvollmer** [comment]
Nested forms don't have a grey background:

image

Previously:

image

**Comment by mvollmer** [comment]
Encryption keys table is not wide enough:

image

Previously:

image

**Comment by mvollmer** [comment]
Field explanations have disappeared:

image

Previously:

image

**Comment by mvollmer** [comment]
Checkbox groups without label are too far right:

image

Previously:

image

**Comment by mvollmer** [comment]
Direct action buttons and the dropdown are too close together:

image

Previously:

image

**Comment by KKoukiou** [comment]
> Dropdowns have a blue focus rings:
> 
> image
> 
> but sometimes a blue focus line:
> 
> image

The second variant is the expected. I will check you example.

**Comment by mvollmer** [comment]
I found two issues that also exist on master.

The "widest title" is visible (Ouch!):

image

And the alert in "Edit NFS" has no margin:

image

I'll fix those separately.

**Review by mvollmer — CHANGES_REQUESTED** [review]

File: pkg/storaged/storage-controls.jsx
Comment (KKoukiou) [reply pkg/storaged/storage-controls.jsx:217]: Done

**Comment by mvollmer** [comment]
> There is now a blue border all around a focused text input field. Is that intentional?
> [kkoukiou]: This is expected, all pF4 text inputs are blue when focused.

Indeed! (I don't like it, but *shrug*)

**Comment by mvollmer** [comment]
> There is a bigger gap between the label and the input elements, which wastes some space.

The gap is still there, right?

**Comment by mvollmer** [comment]
> Text field of size sliders look differently now when focused:
> 
> image
>  
> [kkoukiou]: Expected change

But they look differently than other text input fields, like the VDO name input. Do we haev two kinds now?

**Comment by KKoukiou** [comment]
> There is a b

They are already compact, the spacing is what you see here are well if you specify isCompact, https://github.com/cockpit-project/cockpit/pull/15330

**Comment by KKoukiou** [comment]
> > Text field of size sliders look differently now when focused:
> > image
> > [kkoukiou]: Expected change
> 
> But they look differently than other text input fields, like the VDO name input. Do we have two kinds now?

Ups right, that one I forgot to make PF4.

**Comment by mvollmer** [comment]
> I don't think we should show grey background in nested forms. I even consider it a bug. If you disagree let's ask a designer.

It was designed with a grey background when we first introduced nested forms.  Personally, I think moving away from that because of technology limitations would be a bit weak.

Can we just keep using our old `ct-formt` layout system from `lib/form-layout.scss`?

**Comment by KKoukiou** [comment]
> > I don't think we should show grey background in nested forms. I even consider it a bug. If you disagree let's ask a designer.
> 
> It was designed with a grey background when we first introduced nested forms. Personally, I think moving away from that because of technology limitations would be a bit weak.
> 
> Can we just keep using our old `ct-formt` layout system from `lib/form-layout.scss`?

I did not remove it because of technology limitations, I just  did not explicitly add it when I moved the dialogs to PF4 forms. In cockpit-machines we also have nested forms and we don't have or ever had grey background. form-layout is not meant to be used, was just a solution for when PF4 was not around. If we need some extra feature that PF4 does not offer we should report a RFE upstream and then it's ok to have a custom component. Since I see the design is a bit controversial here, let's just consult @garrett.

**Comment by garrett** [comment]
> It was designed with a grey background when we first introduced nested forms. 

Andreas introduced that in ct-form.

The nested form should have labels on top, not labels on the side. This would probably help to resolve the need for having it styled with a grey background. (However, it might be a little weird on smaller screen sizes?)

I don't think PF designs for complex forms like this. Their progressive disclosure example doesn't even indent at all: https://www.patternfly.org/v4/components/form/design-guidelines/#use-progressive-disclosure

**Comment by mvollmer** [comment]
> In cockpit-machines we also have nested forms and we don't have or ever had grey background.

I see, thanks! Nested forms with a greay background have a problem with disabled elements, which are also grey and thus blend into the background. So if removing the grey background here makes Cockpit more consistent, I am all for it.

**Comment by garrett** [comment]
> removing the grey background here

Please note that more things need to be done than just removing the background. :wink:

**Comment by mvollmer** [comment]
Checkboxes in a nested form have different row spacing than checkboxes in a normal form:

image

**Comment by mvollmer** [comment]
> > There is a b
> 
> They are already compact, the spacing is what you see here are well if you specify isCompact, #15330

Ok, thanks!

**Comment by KKoukiou** [comment]
> > removing the grey background here
> 
> Please note that more things need to be done than just removing the background. wink

Right but let's not make the other things part of this PR please.

**Comment by KKoukiou** [comment]
> Checkboxes in a nested form have different row spacing than checkboxes in a normal form:
> 
> image

Right this is also expected. The first checkboxes, about encryption are form fields, and they don't assosiate to a label. The second group about mount options, are associated to a single label and belong to a group. 

See checkboxes in  https://www.patternfly.org/v4/components/form#horizontal vs https://www.patternfly.org/v4/components/form#horizontal-no-padding-top

**Comment by garrett** [comment]
The labels should be on top in nested forms (as stated above). This can be done either with ct-form or in PF4.

The checkboxes should also have labels, as @KKoukiou says, as they're not part of the password confirmation, but encryption options.

**Comment by KKoukiou** [comment]
> The labels should be on top in nested forms (as stated above). This can be done either with ct-form or in PF4.
> 
> The checkboxes should also have labels, as @KKoukiou says, as they're not part of the password confirmation, but encryption options.

PF4 decided if a form is horizontal or not by pf-m-horizontal class added to the `form` element. That being said, you can't really make some options horizontal and some vertical, because nested forms are not really nested `form` elements, since that would be wrong. So what you ask here, needs more thinking of how it can implemented, and actually I would rather get genuine support from it from PF4 upstream.

File: pkg/storaged/format-dialog.jsx
Comment (mvollmer) [inline pkg/storaged/format-dialog.jsx:75]: I don't understand this, I am afraid. The old code rturned just one `CheckBoxes` element, the new code returns multiple, one for each actual checkbox input.  Why does this need to be changed?

Comment (KKoukiou) [reply pkg/storaged/format-dialog.jsx:75]: its' the same difference you noticed here. https://github.com/cockpit-project/cockpit/pull/15261#issuecomment-779763836

Checkboxes can either belong to one label a a group, and in that case they should be rendered inside a `FromGroup` component. However, Checkboxes often are not part of one label, how the crypto options are, in this case they should be rendered one by one, each wrapped by a `FormGroup` element.

Comment (mvollmer) [reply pkg/storaged/format-dialog.jsx:75]: They belong to a group (except "Store passphrase"), but they didn't have a label. What about this:

image

**Comment by mvollmer** [comment]
The little "danger" icon in a content row is now black:

image

File: pkg/storaged/format-dialog.jsx
Comment (KKoukiou) [reply pkg/storaged/format-dialog.jsx:75]: I am happy with this actually! Makes sense!

**Comment by mvollmer** [comment]
The spinner in the format dialog does not appear anymore:

image

File: pkg/storaged/format-dialog.jsx
Comment (mvollmer) [reply pkg/storaged/format-dialog.jsx]: This change is also responsible for at least one test failure.

**Comment by garrett** [comment]
Something more like this, perhaps?

storage-format-2021-02-16-1322

(I made this with Excalidraw; if you open the PNG with it, it'll let you edit it, as I embedded the data.)

**Comment by KKoukiou** [comment]
> Something more like this, perhaps?
> 
> storage-format-2021-02-16-1322
> 
> (I made this with Excalidraw; if you open the PNG with it, it'll let you edit it, as I embedded the data. Or you can click this link to open it quickly: https://excalidraw.com/#json=6516990898339840,9-oKXT0E41-LgGi0cZG0uA (not sure how long it lasts).)

@garrett with the current PF4 compoent this is not really possible, I commented the reason above, https://github.com/cockpit-project/cockpit/pull/15261#issuecomment-779779979

I can take a look if this is possible to support upstream, but I think it should not block this PR.

**Comment by garrett** [comment]
@KKoukiou: Oh, I missed that comment.

It's possible with ct-form, BTW. Also, it should be made possible in PF4 on a `pf-c-form__group` and/or `pf-c-input-group`.

Meanwhile, the form should collapse down to something like this on mobile view:

storage-format-mobile-2021-02-16-1322

(If PF4 isn't collapsing down, that's another issue.)

**Comment by garrett** [comment]
Moved redesign to #15348.

If we can pick up any low-hanging fruit from the suggestions here, great. Otherwise, it should be done in a follow-up to address those issues.

**Comment by mvollmer** [comment]
The hamburger should not be visible with limited access:

image

**Comment by garrett** [comment]
What are labels aligned to? This is clearly a bug. 

Is it our bug or a PF4 one?

Screenshot from 2021-02-17 13-00-25

Screenshot from 2021-02-17 13-00-16

They're not aligned to the baseline, the middle, the bottom, or even the top of the text or widget.

I think this might be a PF4 bug in some recent version? I'll investigate.

**Comment by garrett** [comment]
It looks like a PF4 bug with alignment and horizontal form elements... :frowning_face: 

Screenshot from 2021-02-17 13-08-19
Screenshot from 2021-02-17 13-07-27

I guess I'll craft an override and file a bug. I still need to finish our override for checkboxes (and probably radio buttons too) and file a bug for that as well. They really need to get their act together with regard to alignment. :disappointed:

**Comment by garrett** [comment]
Filed https://github.com/patternfly/patternfly/issues/3868... I'm looking at a local override solution for us meanwhile.

**Comment by mvollmer** [comment]
I'd say the triangle and spinner on the overview are now too small:

image

Let's make them medium:

image

**Comment by mvollmer** [comment]
Jobs are no longer full width:

image

**Comment by mvollmer** [comment]
The spinner in the content row is too small:

image

**Comment by mvollmer** [comment]
The last column in the content row should be right aligned:

image

image

**Comment by mvollmer** [comment]
How did I end up with two different sized triangles?

image

**Comment by mvollmer** [comment]
> How did I end up with two different sized triangles?

There is a `flex-shrink: 1` active for that flex item, and the broswer will happily make the item smaller, and then for whatever reason the SVG scales itself down.

**Comment by garrett** [comment]
> There is a flex-shrink: 1 active for that flex item, and the browser will happily make the item smaller, and then for whatever reason the SVG scales itself down.

If there's a flex shrinking in effect for that element, then it will shrink when it needs to. As SVG is scalable (that's the S in SVG), it'll rescale (in many circumstances, although that can be changed in some cases—but it's good for icons to be able to scale).

I'll take a look in a bit.

**Comment by mvollmer** [comment]
> I'll take a look in a bit.

Please look at #5360 first, it changes this area drastically.

**Comment by mvollmer** [comment]
The journal has lost the markup for repated entries:

image

Previously:

image

**Comment by mvollmer** [comment]
Spinner and progress text are too close together in dialogs:

image

Previously:

image

**Comment by mvollmer** [comment]
"Encryption options" dialog repeats itself:

image

**Comment by garrett** [comment]
- padding on the right side of the action items in the header is incorrect (it should be smaller)
- padding all around the charts is huge; it should use similar padding to everything else
- padding around the cards' headers seems a bit large as well (most noticable at the top, but also underneith... to the left is OK however and to the right is not really visible due to the whitespace)
- we still have a problem with wrapping (but I think there's a long running PR about this, as well as an upstream PF issue, IIRC) — although I think we should probably just use gap in flex (which doesn't work for WebKit for the immediate future, but it'll finally be fixed in the next version https://caniuse.com/flexbox-gap)
- tables, especially those without headings should probably have a divider line
- filesystems table wraps on the size
- filesystems table has chunky size bars (they could also be made less wide)
- empty tables, such as NFS mounts in this screenshot, should not have table headings

Screenshot_2021-02-22 Storage - admin localhost localdomain(2)

 - many of the same issues above
 - what's "Create new logical volume" supposed to be? A button?
 - The layout falls apart at this size. I'm wondering if we really want to split logical and physical volumes to be side by side. We should definitely not around this size; it becomes debatable at larger widths.

Screenshot_2021-02-22 Storage - admin localhost localdomain(1)

- Same issues as above (makes sense, as this is the same screen at a smaller size)
- Having s smaller width causes lines to show up for the list in the logical volumes. The bottom one should not be there when used in a table like this.
- UUID and Capacity  have a lot of whitespace
Screenshot_2021-02-22 Storage - admin localhost localdomain

**Comment by garrett** [comment]
What I was testing was this branch, but rebased, in a VM. I'm getting different results without it rebased and not in a VM. Hmm... odd. (Many of the same things still apply, but not all, oddly.)

**Comment by KKoukiou** [comment]
> The journal has lost the markup for repated entries:
> 
> image
> 
> Previously:
> 
> image

Hm, so our current, pkg/lib/cockpit-component-logs.jsx needs PF3 to show proper styles. I will need to port it also to PF4. So, this is getting really big, I suggest I reimport PF3, to make the logs work, fix the last issues and we can remove the PF3 inclusion in followups.

**Comment by garrett** [comment]
Many of the issues I pointed out must have been due to a mix and match of compilation of different branches, due to the build system issues with CSS. Sorry for the noise. Happy that the "issues" have been "resolved".

File: pkg/lib/patternfly/patternfly-4-overrides.scss
Comment (garrett) [inline pkg/lib/patternfly/patternfly-4-overrides.scss:74]:
```suggestion
      button:not(.pf-c-dropdown__toggle) {
```

Suuuuuuper minor: This should have a space before the curly bracket.

**Review by garrett — APPROVED** [review]
Looks great now. :+1: 

I spotted a few minor issues that I need to address in other issues/PRs, but nothing with this PR anymore, now that I have it properly built. :blush: 

One very minor thing: There should be a space before the curly bracket. It doesn't do anything functionally nor should it change the CSS that gets compiled and sent to a browser. Therefore, I'm approving this and asking that you optionally consider the change, but no big deal otherwise.

**Comment by garrett** [comment]
Should the NFS mounts have the empty state in the table? Or should it be in place of the table?

PF has an extra small empty state to be used in cards, like this: https://marvelapp.com/prototype/7c5cdhd/screen/72928590

Here's the important part of the mockup that I'm referencing: 
Screenshot_2021-02-23 x-small empty state(1)

Anyway, the point I'm making is that there isn't a set of length 0, which is what the in-table empty state is meant for, but it's something that's not defined. It's like `null` right now and we probably want something more like `undefined`. 

In friendlier words: It's the difference between doing a search (or filter) and finding nothing versus something not yet exiting.

This is related to the PR, but could be resolved in a follow-up, since it's just slight semantics and a little bit of style.

**Comment by KKoukiou** [comment]
> Should the NFS mounts have the empty state in the table? Or should it be in place of the table?
> 
> PF has an extra small empty state to be used in cards, like this: https://marvelapp.com/prototype/7c5cdhd/screen/72928590
> 
> Here's the important part of the mockup that I'm referencing:
> Screenshot_2021-02-23 x-small empty state(1)
> 
> Anyway, the point I'm making is that there isn't a set of length 0, which is what the in-table empty state is meant for, but it's something that's not defined. It's like `null` right now and we probably want something more like `undefined`.
> 
> In friendlier words: It's the difference between doing a search (or filter) and finding nothing versus something not yet exiting.
> 
> This is related to the PR, but could be resolved in a follow-up, since it's just slight semantics and a little bit of style.

I agree, and this pattern exists in other places in cockpit. See machines snapshots. But it's not related to this PR. The NFS tables did not change from master at all here.

File: pkg/lib/patternfly/patternfly-4-overrides.scss
Comment (KKoukiou) [reply pkg/lib/patternfly/patternfly-4-overrides.scss]: Done

**Comment by garrett** [comment]
> I agree, and this pattern exists in other places in cockpit. See machines snapshots. But it's not related to this PR. The NFS tables did not change from master at all here.

It didn't have a heading with an empty state before. If it's like that in master now, then I guess I missed previous commit at some point. Sigh. OK, that's a follow-up then.

**Review by garrett — DISMISSED** [review]
This PR looks great to me!

**Review by mvollmer — DISMISSED** [review]
Uff, thanks!

**Comment by KKoukiou** [comment]
Oh the kdump failure is real, because I changed the spinner the PF4 in the dialog.jsx file. I will repost shortly.

**Review by mvollmer — APPROVED** [review]

**Review by martinpitt — APPROVED** [review]

**Pull request merged**
